### DISCUSSION TITLE
Guide on using new models?

### DISCUSSION THREAD

**Discussion by pylofe** [Crivella/ocr_translate#15] [Ideas]
Hey, so this isnt really a technical issue rather more on the models themselves.

As you mentioned you have enabled the usage of adding your own models, as stated in the readme 
[Supported text OCR models]: 
Hugging Face [Transformers]VisionEncoderDecoder models
With the use of your https://github.com/Crivella/ocr_translate-hugging_face plug in.

As well as in the changelog:
Plugins can now be used to also add models to the database via the following entrypoints:
ocr_translate.box_data
ocr_translate.ocr_data
ocr_translate.tsl_data

I found a hugging face model https://huggingface.co/team-lucid/trocr-small-korean to try adding to the ocr model, however I am quite lost. 

I looked around and landed on the ocr hugging face model in the virt env folder, and the inti_py and the plugins.py, and the entry points.txt in the hugginface dist folder seems like what you mentioned. However I was not able to decide what to do next.

I am wondering if its possible for you to make a rough guide for people with limited technical knowledge to follow in order to use their own models?
Just enough to be used as a blueprint when using more models?

**Comment by pylofe** [comment]
so far what I got to is adding this to the init py in the plug in
team_lucid_ocr_model_data = {
    'name': 'team-lucid/trocr-small-korean',
    'lang': ['ko'],  # Korean language
    'lang_code': 'iso1',
    'entrypoint': 'hugginface.ved'
}

as well as adding 
ocr.hugginface.ved.team_lucid=ocr_translate_hugging_face:team_lucid_ocr_model_data
to the txt file.

**Comment by Crivella** [comment]
# Using entrypoints 

You need to download the [plugin itself](https://github.com/Crivella/ocr_translate-hugging_face/tree/master) and add that dictionary to the `__init__.py` and that line in the [pyproject.toml at this line](https://github.com/Crivella/ocr_translate-hugging_face/blob/af57c9f924aa99a8aa135e6a42cdfd8d2f4be237/pyproject.toml#L56)

Than you can install your own package by doing `pip install .` inside the folder with the pyproject.toml

After you need to run the server with `AUTOCREATE_VALIDATED_MODELS="true"` atleast once for the model to be added in the database 

# Using Django admin page

Another way to just add a model would be to create a superuser in the django project   `python .\manage.py createsuperuser` and than login in the server admin page `127.0.0.1:4000/admin`  from your browser and add it from there (specifying the correct entrypoint to link it to the plugin)

Anyway in the future i plan to add a proper plugin manager, but i'm gonna change the way I am packaging the core right now as `pyinstaller`  does not seem to work with pip

In both cases you need to run the server with `AUTOCREATE_VALIDATED_MODELS="true"` atleast once for the model to be added in the database from the en

## PS
I just tried the mode myself and from the tool they provide on huggingface to test it it seems to work well when you pass images line by line, but not with multiple lines. (Indeed it does not give good results with the tool)
Right now the way this package work, it passes the entire textbox to the model (this could be either improved in the existing VED class, or another one could be written which does the line-splitting beforehand).
This is something I am planning with the future, but will probably try to train some models myself before

**Reply by pylofe** [reply]
So what i did was add the 
team_lucid_ocr_model_data = {
'name': 'team-lucid/trocr-small-korean',
'lang': ['ko'], # Korean language
'lang_code': 'iso1',
'entrypoint': 'hugginface.ved'
}
to the plugins __init__.py as well as adding that ocr.hugginface.ved.team_lucid... line in the [pyproject.toml at this line].
then i did not do the pip install and instead did AUTOCREATE_VALIDATED_MODELS="true", and after trying it , it installed the package on the own, shown via the cmd. 
Not sure if this is how its intended.

But as you stated, yes the model performs very poorly. I tested it and the detection is much much worse then tesseract. 
I'll look into other alternative models if any exist for this language.

**Reply by Crivella** [reply]
The model iteself is not bad, it is just it's not suited for the pipeline i've implemented.
Actually a better way to improve this would be to do the box detection and at that point save also all char position and than let the OCR model decide what to pick from it.
Gonna have to find the time to implement and test this ;P

Actually adding that to the txt the way you did it beofre might've done the trick as that is what pip does when it detects entrypoint in the pyproject (a few minor version before i think it worked differently)

**Comment by Crivella** [comment]
### Adding models (version >=0.5)

Since with the new release i improved the admin interface i guess this question no deserves a new answer:

- Create a superuser
  - f you are using the release file, a super user is already created for you with credentials `admin:password` (do change this if you ever plan to open the server to more than localhost)
  - Otherwise inside the folder with the `manage.py` script you can run `python manage.py createsuperuser` and follow the prompts
- Launch the server
- Connect to the admin interface `http://127.0.0.1:4000/admin`
- On the left click `Add` for the type of models you wish to add
- If you are using the release file, a super user is already
- Fill in the fields appropriately
  - name: the name of the model (should be unique from other models)
  - entrypoint: the entrypoint for the plugin that supports that model. (You can only add models of a type that is already implemented like this, if you want to add a non supported type of model you need to write a plugin for it, or bring it to my attention)
  - Language format (in the future this will be made into a choice field). Useful for models that need the language to be passed to them in some format. Write/select one of the following
    - iso1: Models want language to be written in ISO-639-1 format
    - iso2b: Models want language to be written in ISO-639-2 bibliographic format
    - iso2b: Models want language to be written in ISO-639-2 terminology format
    - iso3: Models want language to be written in ISO-639-3 format   
- iso1_map: For leaving empty just write `{"":""}`. Otherwise used to add non-standard codes that the model requires. The map should be a dictionary that maps an iso1 language code to the one that the model wants (eg `{"zht": "zh"}`)
- default_options: Default options that should be passed to the full pipeline for this model (if in doubt choose `{}` for the defaults implemented in the tool)
- Languages: select the languages this model can be used with
- ocr_mode (for OCR models only): If the model should be passed `Merged` textboxes or `Single` ones (select single for models that are supposed to work only in single line mode. (Single should work in general but will likely be slower for models that support OCR of multi-line text)

![image](https://github.com/Crivella/ocr_translate/assets/34096612/58f3fd64-017b-419d-9564-5aaa79f1e650)

### Through plugins

There is still the option to write a plugin (even without implementing new models) that can act as a collection of models.
For this i would reference the [page in the documentation](https://crivella.github.io/ocr_translate/contrib/plugins.html)
